routing/index_router.cpp

```diff
@@ -374,6 +374,59 @@ IRouter::ResultCode IndexRouter::CalculateRoute(Checkpoints const & checkpoints,
   }
 }
 
+bool IndexRouter::FindBestSegment(m2::PointD const & point, m2::PointD const & direction,
+                                  bool isOutgoing, WorldGraph & worldGraph, Segment & bestSegment,
+                                  bool & bestSegmentIsAlmostCodirectional) const
+{
+  auto const file = platform::CountryFile(m_countryFileFn(point));
+  MwmSet::MwmHandle handle = m_index.GetMwmHandleByCountryFile(file);
+  if (!handle.IsAlive())
+    MYTHROW(RoutingException, ("Can't get mwm handle for", file));
+
+  auto const mwmId = MwmSet::MwmId(handle.GetInfo());
+  NumMwmId const numMwmId = m_numMwmIds->GetId(file);
+
+  vector<pair<Edge, Junction>> candidates;
+  m_roadGraph.FindClosestEdges(point, kMaxRoadCandidates, candidates);
+
+  auto const getSegmentByEdge = [&numMwmId](Edge const & edge) {
+    return Segment(numMwmId, edge.GetFeatureId().m_index, edge.GetSegId(), edge.IsForward());
+  };
+
+  // Getting rid of knowingly bad candidates.
+  my::EraseIf(candidates, [&](pair<Edge, Junction> const & p){
+    Edge const & edge = p.first;
+    return edge.GetFeatureId().m_mwmId != mwmId || IsDeadEnd(getSegmentByEdge(edge), isOutgoing, worldGraph);
+  });
+
+  if (candidates.empty())
+    return false;
+
+  BestEdgeComparator bestEdgeComparator(point, direction);
+  Edge bestEdge = candidates[0].first;
+  for (size_t i = 1; i < candidates.size(); ++i)
+  {
+    Edge const & edge = candidates[i].first;
+    if (bestEdgeComparator.Compare(edge, bestEdge) < 0)
+      bestEdge = edge;
+  }
+
+  bestSegmentIsAlmostCodirectional =
+      bestEdgeComparator.IsDirectionValid() && bestEdgeComparator.IsAlmostCodirectional(bestEdge);
+  bestSegment = getSegmentByEdge(bestEdge);
+  return true;
+}
+
+unique_ptr<WorldGraph> IndexRouter::MakeWorldGraph()
```

Comment:
Am I right that resulting WorldGraph is created from the same params this IndexRouter was constructed from? If so, it's quite strange to have MakeWorldGraph method here.

Reply:
> Am I right that resulting WorldGraph is created from the same params this IndexRouter was constructed from?

Yes. It makes SingleVehicleWorldGraph (derived from WorldGraph) with params which kept at IndexRouter.

I moved to public section, because I need to use method IndexRouter::FindBestSegment(). 

> If so, it's quite strange to have MakeWorldGraph method here.

Could you please write in more details what you mean?

Reply:
Странно что у индекс роутера есть публичный метод для создания WorldGraph, который внутри создает WorldGraph примерно из того же набора параметров, из которого был создан IndexRouter.
Кажется что если у вызывающего кода есть все эти параметры для создания IndexRouter, то и WorldGraph он может создать без посредничества IndexRouter.

Reply:
Как вариант, можно добавить публичный метод IndexRouter::FindBestSegment(), с меньшим числом параметров, и в нем создавать WorldGraph и т.п.

Reply:
+1 к @tatiana-kondakova, именно это и имел ввиду в своем первом комменте. Так ли нужно создавать WorldGraph внутри IndexRouter, или он может приходить извне? И если может, станет ли проще код?

Reply:
вернуть как было и сделать публичный метод, вызывающий MakeWorldGraph и FindBestSegment?

Reply:
> вернуть как было и сделать публичный метод, вызывающий MakeWorldGraph и FindBestSegment?

Давайте так и поступим.

Reply:
Готово